File: credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int main(void)
{
    long card = get_long("Card number: ");
   
    if ((card / (pow(10, 12))) == 0) //only card numbers
    {
        printf("INVALID\n");
    }

    else
    {

        int fsum = 0;
        long aux = 0;
    
        for (int i = 1; i < 16; i += 2)
        {
            aux = (card / pow(10, (i)));

            if ((((aux % 10) * 2) / 10) == 0)
            {
                fsum = fsum + ((aux % 10) * 2); //sum with one digit
            }
            else
            {
                //separete sum for 2 digits
                fsum = fsum + (((aux % 10) * 2) / 10);
                fsum = fsum + (((aux % 10) * 2) % 10);
            }
        }
    
        for (int j = 0; j < 16; j += 2)
        {
            aux = (card / pow(10, (j)));
            fsum = fsum + (aux % 10); //sum the remaining digits
        }

        int amex, visa13, visa16, master; //auxs to make comparation easier

        amex = (card / (pow(10, 13)));
        visa13 = (card / (pow(10, 12)));
        visa16 = (card / (pow(10, 15)));
        master = (card / (pow(10, 14)));

        if ((fsum % 10) != 0)
        {
            printf("INVALID\n");
        }

        else if (amex == 34 || amex == 37) //if amex
        {
            printf("AMEX\n");
        }

        else if (visa13 == 4 || visa16 == 4) //if visa
        {
            printf("VISA\n");
        }

        else if (master >= 51 && master <= 55) //if mastercard
        {
            printf("MASTERCARD\n");
        }

        else
        {
            printf("INVALID\n");
        }

    } 
}

File: credit.py
from cs50 import get_int
from sys import exit

# user card number
card = get_int("Number: ")

# condition for only card numbers
if card // pow(10, 12) == 0:
    print("INVALID")
    exit(1)

# Luhn's Algorithm
sum = 0

# 1st step = x2 other and sum digits
for i in range(1, 16, 2):

    left = card // pow(10, i)
    digit = left % 10

    if digit * 2 < 10:
        sum += digit * 2
    else:
        sum += digit * 2 // 10
        sum += digit * 2 % 10

# 2nd step = sum to the sum of non multiplied digits
for j in range(0, 16, 2):

    left = card // pow(10, j)
    sum += left % 10

# 3rd step = if the last digit of sum is 0, the card is valid
if not sum % 10 == 0:
    print("INVALID")
    exit(1)

# if AMEX
if card // pow(10, 13) == 34 or card // pow(10, 13) == 37:
    print("AMEX")
    exit(0)

# if VISA
if card // pow(10, 12) == 4 or card // pow(10, 15) == 4:
    print("VISA")
    exit(0)

# if MASTER
if card // pow(10, 14) >= 51 and card // pow(10, 14) <= 55:
    print("MASTER")
    exit(0)

print("INVALID")

